### PR TITLE
Update libsecp256k1

### PR DESCRIPTION
**Pull request by sipa**
Libsecp256k1 now has explicit context objects, which makes it completely thread-safe. In turn, keep an explicit context object in key.cpp, which is explicitly initialized destroyed. This is not really pretty now, but it's more efficient than the static initialized object in key.cpp (which made for example bitcoin-tx slow, as for most of its calls, libsecp256k1 wasn't actually needed).

This also brings in the new blinding support in libsecp256k1. By passing in a random seed, temporary variables during the elliptic curve computations are altered, in such a way that if an attacker does not know the blind, observing the internal operations leaks even less information about the keys used. This was implemented by Greg Maxwell.

### CODE REVIEW

**Comment by jtimon** [comment]
Random question would it make sense for libsecp256k1 a submodule (http://git-scm.com/book/en/v2/Git-Tools-Submodules) of bitcoin core?
I've never used them but it seems to me that this could be a case where they make sense.

**Comment by sipa** [comment]
There is a long debate between users of submodules and of subtrees. I'd
rather not repeat it here :)

**Comment by jtimon** [comment]
Oh, I see. But I assume in principle any of those would serve...shirt or t-shirt? Though choice, better go out naked.

**Comment by sipa** [comment]
We *are* using git subtree (for libsecp256k1 and for leveldb). This is a
pull request with a merge made by git subtree. There is a script in contrib
or devtools that can verify that the subdirectory matches the history of
the external repo.

**Comment by jtimon** [comment]
Oh I see, great. Never mind then, thanks for explaining.

File: src/bitcoin-tx.cpp [hunk 1]
```diff
@@ -475,6 +478,8 @@ static void MutateTx(CMutableTransaction& tx, const string& command,
 
     else
         throw runtime_error("unknown command");
+
+    ECC_Stop();
```
Comment (laanwj) [inline src/bitcoin-tx.cpp]: Let's use a RAII approach here. Otherwise, there will be a leak when control flow leaves the function through an exception.

**Comment by laanwj** [comment]
utACK

**Pull request merged**